libevmasm/Assembly.cpp

```diff
@@ -478,8 +478,15 @@ Json::Value Assembly::assemblyJSON(std::map<std::string, unsigned> const& _sourc
 	{
 		root["sourceList"] = Json::arrayValue;
 		Json::Value& jsonSourceList = root["sourceList"];
-		for (auto const& [name, index]: _sourceIndices)
-			jsonSourceList[index] = name;
+		unsigned maxSourceIndex{};
+		for (auto const& [sourceName, sourceIndex]: _sourceIndices)
+		{
+			maxSourceIndex = std::max(sourceIndex, maxSourceIndex);
+			jsonSourceList[sourceIndex] = sourceName;
+		}
+		for (unsigned i = 0 ; i < maxSourceIndex; ++i)
+			if (jsonSourceList[i] == Json::nullValue)
+				jsonSourceList[i] = "unknown-source-" + std::to_string(i);
```

Comment:
I adopted this specific default value because it was the choice made by @aarlt in the original PR (see this commit: https://github.com/ethereum/solidity/commit/136da0c74c678c91fe6a4a5bb36af7ddbfdea376). I'm uncertain if there was a specific reason for this choice.

However, if we were to use `null` instead, distinguishing between the different unknown sources would not be possible. This raises the question of whether the use of `unknown-source-X` was intended to provide a unique identifier, and if so, whether this differentiation is necessary.

Reply:
In general it was needed to have a valid index in `jsonSourceList`, but it could have been sufficient to just allow `null` here. 

Sorry for the late answer!

Reply:
I don't remember a specific reason why I set this to `unknown-source-X`. I think `null` is enough and somehow I just didnt saw that option when I implemented it initally. I guess I thought that the `null` would be removed - but of course it makes no sense within an array.

Reply:
Ok, I did this change. I just created a very simple test that is importing such assembly again. This test - of course- failed, because we don't have proper support for `null` in source indices. Initially I thought that the changes supporting `null` are somehow trivial - but it turned out that we need to change many files to support this change. I would say,  we should just go with `unknown-source-X`, but probably name that `#unknown-source-X`. I guess the prepended `#` shows that it is not a real file, like we have it for ` #utility.yul`.

Reply:
In that case I think it's best to just have it fail on import. The cases where a `null` would be produced are really ones where the debug data is incomplete. The compiler will never produce such broken debug annotations when generating Yul so we don't really have much need to support importing them. Might even be a good idea to fail already when such Yul code is supplied, but that would be breaking so we can't do that. Failing on asm import is not breaking though (in fact we already do that on `null`).

I'd rather insert `null` and fail fast than try hard and pretend everything is fine to the point of inventing fake source names.

Reply:
Youre totally right! I don't know - somehow I didn't saw that option to just fail :D I added a new tests `asm_json_import_sourcelist_with_null_elements` and changed `asm_json_export_yul_arbitrary_source_index`. Both tests are failing now with an error.

Reply:
Ok, finally I had another look to this PR. It turned out that it is possible that the debug data is incomplete. Just take a look at the output of e.g. `solc/solc ../test/externalTests/solc-js/DAO/TokenCreation.sol --ir-optimized --debug-info location  --optimize`. You will see that a yul object is generated, that will only have a source map starting from `1` - `/// @use-src 1:"/Users/alex/git/solidity/test/externalTests/solc-js/DAO/Token.sol"` - without defining a source for index `0`. This is due to the fact that the IR Generator will only output the debug information for used files. So I just removed the `null` check in `Assembly::assemblyJSON(..)`. This prevented the via-ir-equivalence tests from passing. However, for importing I left the null check in `Assembly::fromJSON(..)` - it will only be possible to import assembly json with fully specified source lists.

Reply:
I just did another look to this, because I was wondering why exactly in the context of the via-ir-equivalence tests assembly json code was executed. It turned out that it will just always calculate the assembly json object, that is not really needed. So I just changed that now - and because the assembly json object gets only created upon request, I put the `null` check back to `Assembly::assemblyJSON(..)`.

Reply:
Interesting. I did not even realize we sometimes do generate non-contiguous source indices.

Well, the last thing to check here then is it possible for us to run into this check in `Assembly::assemblyJSON()` when compiling Solidity source and requesting asm export? If it is, then that would be breaking...

Reply:
I just did a small local test with the contract mentioned above: `solc/solc ../test/externalTests/solc-js/DAO/TokenCreation.sol --ir-optimized --debug-info location  --optimize --asm-json` is working correctly.

Reply:
Great. I can confirm that it's fine - apparently assembly export always gets the full list of sources from `CompilerStack` an lists them all so there are no gaps. Unlike Yul, which only includes sources that are actually referenced and may have gaps.

Actually, this looks like a small bug in assembly export, I think that `--asm-json` should behave like Yul in that regard. Otherwise including extra files in compilation results in different assembly output. But this was not introduced by this PR so it's out of scope.